server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -2199,10 +2208,16 @@ private Set<DataSegmentPlus> createNewIdsOfAppendSegmentsAfterReplace(
           newInterval = replaceInterval;
           break;
         } else if (replaceInterval.overlaps(oldInterval)) {
-          throw new ISE(
-              "Incompatible segment intervals for commit: [%s] and [%s].",
-              oldInterval, replaceInterval
-          );
+          final String conflictingSegmentId = oldSegment.getId().toString();
+          final String upgradeVersion = upgradeSegmentToLockVersion.get(conflictingSegmentId);
+          throw DruidException.forPersona(DruidException.Persona.OPERATOR)
+                              .ofCategory(DruidException.Category.UNSUPPORTED)
+                              .build(
+                                  "Replacing with a finer segment granularity than a concurrent append is unsupported."
+                                  + " Cannot upgrade segment[%s] to version[%s] as the replace interval[%s]"
+                                  + " does not fully contain the pending segment interval[%s].",
```

Comment:
The interval check here is for overlapping conditions rather than full containment, but the error message states: `"does not fully contain the pending segment interval"`, which is confusing. Is this a copy-paste error, or should the check be aligned with `shouldUpgradePendingSegment()` above, which checks for interval `contains()`?

Reply:
The first if condition is to see if there is a complete containment, failing which we check if there's an overlap (partial)

Reply:
Ah, I see, okay - thanks for the clarification. So this is effectively the same condition as ` if (!replaceInterval.contains(oldInterval)` but written in a different way. Maybe we can clean this up later.